subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/testsuites/TestSuitesDependenciesIntegrationTest.groovy

```diff
@@ -0,0 +1,233 @@
+            import org.junit.Assert;
+            import org.junit.Test;
+
+            public class IntegrationTest {
+                @Test
+                public void integrationTest() {
+                    Assert.assertTrue(StringUtils.isEmpty("")); // compilation also fails here; commons-lang3 is not automatically "inherited" by integTests
+                }
+            }
+        '''
+
+        when:
+        fails 'check'
+
+        then:
+        failureCauseContains('Compilation failed; see the compiler error output for details.')
+    }
+
+    def 'default test suite has project dependency by default; others do not'() {
+        given:
+        buildFile << """
+        plugins {
+          id 'java'
+        }
+
+        ${mavenCentralRepository()}
+
+        dependencies {
+            // production code requires commons-lang3 at runtime, which will leak into tests' runtime classpaths
+            implementation 'org.apache.commons:commons-lang3:3.11'
+        }
+
+        testing {
+            suites {
+                integTest(JvmTestSuite)
+            }
+        }
+
+        tasks.named('check') {
+            dependsOn testing.suites.integTest
+        }
+
+        tasks.register('checkConfiguration') {
+            dependsOn test, integTest
+            doLast {
+                assert configurations.testRuntimeClasspath.files*.name == ['commons-lang3-3.11.jar'] : 'commons-lang3 leaks from the production project dependencies'
+                assert !configurations.integTestRuntimeClasspath.files*.name.contains('commons-lang3-3.11.jar') : 'integTest does not implicitly depend on the production project'
+            }
+        }
+        """
+
+        expect:
+        succeeds 'checkConfiguration'
+    }
+
+    def 'non-default default test suites have project dependency if explicitly set'() {
+        given:
+        buildFile << """
+        plugins {
+          id 'java'
+        }
+
+        ${mavenCentralRepository()}
+
+        dependencies {
+            // production code requires commons-lang3 at runtime, which will leak into tests' runtime classpaths
+            implementation 'org.apache.commons:commons-lang3:3.11'
+        }
+
+        testing {
+            suites {
+                integTest(JvmTestSuite) {
+                    dependencies {
+                        implementation project(':')
+                    }
+                }
+            }
+        }
+
+        tasks.named('check') {
+            dependsOn testing.suites.integTest
+        }
+
+        tasks.register('checkConfiguration') {
+            dependsOn test, integTest
+            doLast {
+                assert configurations.testCompileClasspath.files*.name == ['commons-lang3-3.11.jar'] : 'commons-lang3 leaks from the production project dependencies' // TODO possible bug; why is commons-lang3 present?
```

Comment:
@tresat @big-guy this assertion passes, but should fail IMO. `commons-lang3` is an `implementation` dependency of the root project, and shouldn't be exposed to the compile-time classpath of consumers.  Thoughts?

Reply:
This is a consequence of the relationship between testImplementation and implementation described in the diagram here:
https://docs.gradle.org/current/userguide/java_plugin.html#tab:configurations

I think this is OK because we're trying to treat "test" like a unit test suite, which might have more tendrils into the real code than an integration test. It's been like this forever.

Reply:
I think this is correct, the `TestCompileClasspath` should extend `testImplementation` which extends `implementation` - https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_configurations_graph, right?

Reply:
Thanks, resolved by d5a439e8359